multiapps-mta/src/main/java/org/cloudfoundry/multiapps/mta/resolvers/ReferencesFinder.java

```diff
@@ -0,0 +1,97 @@
+import org.cloudfoundry.multiapps.mta.helpers.VisitableObject;
+import org.cloudfoundry.multiapps.mta.model.DeploymentDescriptor;
+import org.cloudfoundry.multiapps.mta.model.ElementContext;
+import org.cloudfoundry.multiapps.mta.model.Hook;
+import org.cloudfoundry.multiapps.mta.model.Module;
+import org.cloudfoundry.multiapps.mta.model.ParametersContainer;
+import org.cloudfoundry.multiapps.mta.model.PropertiesContainer;
+import org.cloudfoundry.multiapps.mta.model.ProvidedDependency;
+import org.cloudfoundry.multiapps.mta.model.RequiredDependency;
+import org.cloudfoundry.multiapps.mta.model.Resource;
+import org.cloudfoundry.multiapps.mta.model.Visitor;
+
+public class ReferencesFinder extends Visitor implements SimplePropertyVisitor {
+
+    private Set<String> foundReferences;
+
+    public void fillWithReferences(DeploymentDescriptor descriptor, Set<String> referenceContainer) {
+        this.foundReferences = referenceContainer;
+        descriptor.accept(this);
+    }
+
+    @Override
+    public void visit(ElementContext context, DeploymentDescriptor descriptor) {
+        findReferencesInParameters(descriptor);
+    }
+
+    @Override
+    public void visit(ElementContext context, Module module) {
+        findReferencesInProperties(module);
+        findReferencesInParameters(module);
+    }
+
+    @Override
+    public void visit(ElementContext context, ProvidedDependency providedDependency) {
+        findReferencesInProperties(providedDependency);
+        findReferencesInParameters(providedDependency);
+    }
+
+    @Override
+    public void visit(ElementContext context, RequiredDependency requiredDependency) {
+        findReferencesInProperties(requiredDependency);
+        findReferencesInParameters(requiredDependency);
+    }
+
+    @Override
+    public void visit(ElementContext context, Resource resource) {
+        findReferencesInProperties(resource);
+        findReferencesInParameters(resource);
+    }
+
+    @Override
+    public void visit(ElementContext context, Hook hook) {
+        findReferencesInParameters(hook);
+    }
+
+    private void findReferencesInProperties(PropertiesContainer propertiesContainer) {
+        findReferencesInMap(propertiesContainer.getProperties());
+    }
+
+    private void findReferencesInParameters(ParametersContainer parametersContainer) {
+        findReferencesInMap(parametersContainer.getParameters());
+    }
+
+    private void findReferencesInMap(Map<String, Object> map) {
+        Map<String, Object> modifiableMap = new HashMap<>(map);
+        new VisitableObject(modifiableMap).accept(this);
+    }
+
+    @Override
+    public Object visit(String key, String value) {
+        referencePatternMatches(value);
+        return value;
+    }
+
+    private void referencePatternMatches(String valueToMatch) {
+        for (Pattern compiledPattern : ReferencePattern.COMPILED_PATTERNS) {
+            Matcher matcher = compiledPattern.matcher(MiscUtil.cast(valueToMatch));
```

Comment:
each pattern actually has specific uses for the matcher.groups after matching - maybe this should all be handled externally by ReferencePattern, since the 3 different pattern are expected to have or not have different count of groups after a match - ${%s} vs ~{%s} vs ~{%s/%s}
The random iteration over every regex matched group could (maybe?) lead to wrong validation/weird warning messages - e..g 
'\~{not_existing_param1/not_existing_param2/existging_param}'